#1054 - Change  config key parsing to attempt Base64 decoding first.

dmitrif opened the pull request:
## Summary
My GO skills are non-existent but this is my stab to allow `certificates` to first try base64 decoding for the entries, and only then fallback to FS resolution.

## Related issues
Addresses #1054 

**Checklist**:
- [x] add related issues
- [x] updated docs (Bypassed as they are currently ambiguous)
- [x] updated unit tests (Bypassed due to lack of coverage)
- [x] updated UPGRADING.md (Bypassed due to not being a breaking change)
- [x] ready for review

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pomerium/pomerium?pullRequest=1055) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pomerium/pomerium?pullRequest=1055) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pomerium/pomerium?pullRequest=1055) it.</sub>

cuonglm commented:
@dmitrif would you mind adding a test for this?

dmitrif commented:
Thanks! 

@cuonglm / @desimone : I am happy to add a test case for this. However struggling to get the test harness to run locally with `make: opa: No such file or directory`. Is there a contributors guide by chance to get this running? Cheers!

cuonglm commented:
> Thanks!
> 
> @cuonglm / @desimone : I am happy to add a test case for this. However struggling to get the test harness to run locally with `make: opa: No such file or directory`. Is there a contributors guide by chance to get this running? Cheers!

`opa` is required to run rego policy tests, in this case, you can simply do:

```
go test -race $(go list ./... | grep -v vendor | grep -v github.com/pomerium/pomerium/integration)
```

dmitrif commented:
@cuonglm Thank you, that worked perfectly!

@cuonglm / @desimone : Added unit test coverage and updated logging to be a bit more verbose to output both the base64 error and the file parsing error if both scenarios fail.

cuonglm commented on config/options.go:
```diff
@@ -557,11 +557,24 @@ func (o *Options) Validate() error {
-		if err != nil {
-			return fmt.Errorf("config: bad cert file %w", err)
+	for i, c := range o.CertificateFiles {
+		fileRefErr := error(nil)
```
use `var fileRefErr error` instead

cuonglm commented on config/options.go:
```diff
@@ -557,11 +557,24 @@ func (o *Options) Validate() error {
+	for i, c := range o.CertificateFiles {
+		fileRefErr := error(nil)
+		cert, base64Err := cryptutil.CertificateFromBase64(c.CertFile, c.KeyFile)
+		
```
Remove this empty line.

cuonglm commented on config/options.go:
```diff
@@ -557,11 +557,24 @@ func (o *Options) Validate() error {
+		fileRefErr := error(nil)
+		cert, base64Err := cryptutil.CertificateFromBase64(c.CertFile, c.KeyFile)
+		
+		if base64Err != nil {
```
I don't think this need to be complicated like this. Just use one error like your original version, and said that "either a base64 or a file reference" in error message.

cuonglm commented on config/options_test.go:
```diff
@@ -505,6 +505,37 @@ func TestHTTPRedirectAddressStripQuotes(t *testing.T) {
+	tests := []struct {
+		name             string
+		certificateFiles []certificateFilePair
+		wantErr 				 bool
```
`go fmt` this.

cuonglm commented on config/options_test.go:
```diff
@@ -505,6 +505,37 @@ func TestHTTPRedirectAddressStripQuotes(t *testing.T) {
+		{ "Returns an error otherwise", []certificateFilePair{{ KeyFile: "abc", CertFile: "abc" }}, true },
+	}
+
+	for _, tt := range tests {
```
You can run parallel here.

dmitrif commented:
@cuonglm Thank you! Simplified back to the previous iteration and fixed the rest of CR comments.

cuonglm commented on config/options_test.go:
```diff
@@ -505,6 +505,41 @@ func TestHTTPRedirectAddressStripQuotes(t *testing.T) {
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```
You need `tt := tt` before `t.Run`, so it won't capture wrong `tt` variable in for loop.

dmitrif replied:
Will add, but could you please elaborate? I assumed that Go would introduce a new scope when the test func was declared. As such there should be no scope confusion about `tt`?

Thanks!

cuonglm replied:
I invite you to read https://blog.golang.org/subtests, read `Run a group of tests in parallel` section.

Tests run without parallel is safe to not capture the range variable.

dmitrif replied:
Thank you! I made the change, but this is a great starting point for me understanding why :) Appreciate your patience!

cuonglm replied:
> Thank you! I made the change, but this is a great starting point for me understanding why :) Appreciate your patience!

Would you please formatting files with `gofmt -s` 😄
